spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/transaction/ReactiveFirestoreTransactionManager.java

```diff
@@ -0,0 +1,189 @@
+
+			return holder.doOnNext(transactionObject::setResourceHolder)
+					.onErrorMap(
+							ex -> new TransactionSystemException("Could not start Firestore transaction", ex))
+					.doOnSuccess(resourceHolder -> {
+						synchronizationManager.bindResource(getRequiredDatabase(), resourceHolder);
+					})
+					.then();
+		});
+	}
+
+	@Override
+	protected Mono<Void> doCommit(TransactionSynchronizationManager transactionSynchronizationManager,
+			GenericReactiveTransaction genericReactiveTransaction) throws TransactionException {
+		ReactiveFirestoreResourceHolder resourceHolder =
+				extractFirestoreTransaction(genericReactiveTransaction).getResourceHolder();
+
+		CommitRequest.Builder builder = CommitRequest.newBuilder()
+				.setDatabase(this.databasePath)
+				.setTransaction(resourceHolder.getTransactionId());
+
+		resourceHolder.getWrites().forEach(builder::addWrites);
+
+		return ObservableReactiveUtil
+				.<CommitResponse>unaryCall(obs -> this.firestore.commit(builder.build(), obs))
+				.then();
+	}
+
+	@Override
+	protected Mono<Void> doRollback(TransactionSynchronizationManager transactionSynchronizationManager,
+			GenericReactiveTransaction genericReactiveTransaction) throws TransactionException {
+		ReactiveFirestoreTransactionObject transactionObject = extractFirestoreTransaction(genericReactiveTransaction);
+
+		return ObservableReactiveUtil
+				.<Empty>unaryCall(
+						obs -> this.firestore.rollback(RollbackRequest.newBuilder()
+								.setTransaction(transactionObject.getTransactionId()).build(), obs))
+				.then();
+	}
+
+	private FirestoreGrpc.FirestoreStub getRequiredDatabase() {
+		return this.firestore;
+	}
+
+	private static ReactiveFirestoreTransactionObject extractFirestoreTransaction(Object transaction) {
+
+		Assert.isInstanceOf(ReactiveFirestoreTransactionObject.class, transaction,
+				() -> String.format("Expected to find a %s but it turned out to be %s.",
+						ReactiveFirestoreTransactionObject.class,
+						transaction.getClass()));
+
+		return (ReactiveFirestoreTransactionObject) transaction;
+	}
+
+	private static ReactiveFirestoreTransactionObject extractFirestoreTransaction(GenericReactiveTransaction status) {
+
+		Assert.isInstanceOf(ReactiveFirestoreTransactionObject.class, status.getTransaction(),
+				() -> String.format("Expected to find a %s but it turned out to be %s.",
+						ReactiveFirestoreTransactionObject.class,
+						status.getTransaction().getClass()));
+
+		return (ReactiveFirestoreTransactionObject) status.getTransaction();
+	}
+
+	private Mono<ReactiveFirestoreResourceHolder> newResourceHolder(TransactionDefinition definition) {
+		TransactionOptions transactionOptions = definition.isReadOnly()
+				? TransactionOptions.newBuilder().setReadOnly(TransactionOptions.ReadOnly.newBuilder().build()).build()
+				: TransactionOptions.newBuilder().setReadWrite(TransactionOptions.ReadWrite.newBuilder().build())
```

Comment:
You can skip the intermediate `.build()` -- proto setters all have overloaded versions that take builders of subfields.

Reply:
done